src/decompress.cpp

```diff
@@ -81,36 +114,29 @@ auto decompress_block_huffman(
     const huffman::table<std::uint16_t, fixed_dist_table_size>& dist_table)
     -> DecompressStatus
 {
-  std::uint16_t lit_or_len{};
   while (true) {
     const auto lit_or_len_decoded = huffman::decode_one(len_table, src_bits);
     if (not lit_or_len_decoded.encoded_size) {
       return DecompressStatus::InvalidLitOrLen;
     }
-    lit_or_len = lit_or_len_decoded.symbol;
     src_bits.consume(lit_or_len_decoded.encoded_size);
-    if (lit_or_len < detail::lit_or_len_end_of_block) {
-      dst[static_cast<std::size_t>(dst_written++)] =
-          static_cast<std::byte>(lit_or_len);
-      continue;
-    }
-    if (lit_or_len == detail::lit_or_len_end_of_block) {
-      break;
-    }
-    if (lit_or_len > detail::lit_or_len_max) {
+    const auto maybe_lit_or_len =
```

Comment:
This looks better to me.

I think this could still benefit from breaking out smaller functions - although if you think that's too much work, feel free to ignore.

```cpp
  const auto done = 
    parse_lit_or_len(lit_or_len_decoded.symbol, src_bits)
      .and_then(decode_lit_or_len_into(dst))
      .or_else(convert_error_status)
      
   if (done) {
     return *done;
   }

   // done is empty, so the the loop continues
```

where `convert_error_status: ParseLitOrLenStatus -> optional<DecompressStatus>`
```cpp
return {in_place, e == EndOfBlock ? Success : InvalidLitOrLen};
```

and `decode_lit_or_len_into` uses the poor version of pattern matching available in C++:
```cpp
auto decode_lit_or_len_into(std::span<std::byte>& dst)
{
  return [&dst](variant<std::byte, std::uint16_t> lit_or_len) {
    return visit(
      overloaded({
        [&dst](byte lit) { return decode_into(lit, dst); },
        [&dst](uint16_t len) { return decode_into(lit, dst); }
      }),
      lit_or_len
    );
  };
}

// https://en.cppreference.com/w/cpp/utility/variant/visit
template <class... Ts>
struct overloaded : Ts... { using Ts::operator()...; };
```

and

```cpp
auto decode_into(byte lit, span& dst)
  -> optional<DecompressStatus>
{
  // lines 133-137
}
auto decode_into(uint16_t len, span& dst)
  -> optional<DecompressStatus>
{
 // lines 139-161
}
```

although that's a bit of a simplification - you'll also need to capture `dst_written`.

The different cases are broken out into different functions, although there's some logic inversion with `variant::visit`.